DO NOT MERGE! AES256CounterRNG built-in replacement (#12636)

gosubpl opened the pull request:
This PR is intended to discuss the possible replacement of AES256CounterRNG by alternative built-in function doing the same.

patriknw committed:
shard coordinator should wait until min-members regions registered, #21194

wojtasskorcz committed:
Adding create() to ReceiveBuilder to promote a builder-like syntax #18894

gosubpl committed:
AES256CounterRNG built-in replacement (#12636)

gosubpl committed:
AES256CounterRNG built-in replacement (#12636)

gosubpl commented:
Refs #12636

gosubpl commented:
@patriknw @drewhk would be grateful if you took a look at it.
This is an incomplete code - for discussion only - **please do not merge** (yet)

AESCounterBuiltinRNG is a fresh implementation in Scala of the AESCounterRNG algorithm from uncommons.math. The whole library is Apache License 2.0 and the AES CTR RNG is a folk-lore algorithm widely that is widely used.

It is accompanied by AES256CounterBuiltinRNG which is a copy of AES256CounterRNG wrapping the new algorithm (AESCounterBuiltinRNG) instead of the old one (AES256CounterRNG). It also contains direct reference to Java's SecureRandom seed generation instead of going through uncommons.math wrapper.

Two accompanying tests show that:

1. The new code generates the same random numbers as the old one
2. The performance of the new code is similar to old one (I would say around 10-15% worse - see copied example runs of jmh benchmar):
```
akka-bench-jmh > jmh:run -wi 15 -i 15 -f 1 .*RandomReplacementBenchmark.*
[...]
[info] Benchmark                          Mode  Cnt    Score    Error  Units

[info] RandomReplacementBenchmark.runNew    ss   15  254.022 ± 31.966  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  261.415 ± 36.318  ms/op

[info] RandomReplacementBenchmark.runNew    ss   15  372.153 ± 23.552  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  306.477 ± 32.487  ms/op

[info] RandomReplacementBenchmark.runNew    ss   15  249.846 ± 28.165  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  210.568 ± 14.917  ms/op

[info] RandomReplacementBenchmark.runNew    ss   15  373.674 ±  85.931  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  384.629 ± 101.545  ms/op

[info] RandomReplacementBenchmark.runNew    ss   15  249.659 ± 20.295  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  262.815 ± 49.112  ms/op
```

Possible strategies of going forward are:
1. Add AESCounterBuiltinRNG together with AES256CounterBuiltinRNG and AES128CounterBuiltinRNG to codebase, add them also to the list of available codes and deprecate the AES256CounterRNG (together with AES128CounterRNG). To remove uncommons.math at later time remove the the already deprecated Inet implementations and the non-builtin implementations.
2. Replace AESCounterRNG with AESCounterBuiltinRNG in AES256CounterRNG (and its 128 bit counterpart) - this will effectively replace the AES256CounterRNG implementation with the AES256CounterBuiltinRNG implementation. After that to remove uncommons.math just remove the already deprecated Inet RNG. Will be binary incompatible but the protocol will match (as the generated numbers are the same)

I would suggest going forward with 2.

gosubpl committed:
AES key changed to 128 bit to work around Jenkins policy limitations (#12636)

gosubpl committed:
AES key changed to 128 bit to work around Jenkins policy limitations (#12636)

drewhk commented on akka-remote/src/main/scala/akka/remote/security/provider/AESCounterBuiltinRNG.scala:
```diff
@@ -0,0 +1,93 @@
+  private var index: Int = 0
+  private var currentBlock: Array[Byte] = null
+
+  private val cipher = Cipher.getInstance("AES/ECB/NoPadding")
```
wait, doesn't Cipher has an AES/CTR/NoPadding anyway? Why hand-build the counter? CTR just XORs the plaintext with the current counter output. I.e. you should just continuously encrypt the all-zero block and take its output as the random value. Also, I would prefer if this is automatically reseeded from time to time from the SecureRandom, i.e. after a certain number of outputs.

gosubpl replied:
I have added `AESCounterBuiltinCTRRNG` that is using `AES/CTR/NoPadding`. To allow verification I have added custom iv with value of 1 but, obviously, this can be removed before merge.
Works great, only issue being that it is even slower, as the zero block needs now to be XORed with the encrypted counter. So the performance looks like that:
```
[info] Benchmark                          Mode  Cnt    Score    Error  Units
[info] RandomReplacementBenchmark.runCtr    ss   15  299.485 ± 23.466  ms/op
[info] RandomReplacementBenchmark.runNew    ss   15  238.772 ± 12.127  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  215.453 ± 24.637  ms/op

[info] RandomReplacementBenchmark.runCtr    ss   15  283.690 ± 17.258  ms/op
[info] RandomReplacementBenchmark.runNew    ss   15  251.412 ± 25.066  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  231.133 ± 29.139  ms/op

[info] RandomReplacementBenchmark.runCtr    ss   15  314.895 ± 45.711  ms/op
[info] RandomReplacementBenchmark.runNew    ss   15  250.495 ± 15.762  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  215.201 ± 23.651  ms/op
```
which means that the CTR algo is **1.5x** slower than the original one. If that is fine, great, I will happily replace that as hand-built counter is also hurting my eyes.
Of course, aye to reseeding. This is missing from the original implementation, and, for realistic messaging volumes, badly needed. I will add this as soon as we settle the algorithm.

richardimaoka committed:
Add DistributedPubSub message guarantee in doc (Fixes #21441) (#21713)

patriknw committed:
Merge pull request #21715 from wojtasskorcz/wip-mutable-syntax-for-receive-builder

Make ReceiveBuilder a real mutable builder #18894

ktoso committed:
+doc add missing java documentation for scanAsync/foldAsync (#21701)

* +doc add missing java documentation for scanAsync/foldAsync

=doc fix formatting in stages overview (scanAsync/foldAsync)

* Future => CompletionStage

ktoso committed:
+rem remote deployment whitelist feature (#21676)

* +rem remote deployment whitelist feature (initial commit)

+rem remote deployment whitelist feature (initial commit)

* +doc,rem added docs on remoting whitelist

* Update remoting.rst

* Update remoting.rst

ktoso committed:
remove akka parsing (moved to akka/akka-http) (#21721)

ktoso committed:
+str #20795 IOResult construction exposed, in bincompat way (#21070)

* +str #20795 IOResult construction exposed, in bincompat way

Thanks https://github.com/ktoso/kaze-class ;)

* Addressed feedback in IOResult

* Update IOResult.scala

* Update IOResult.scala

gosubpl committed:
another alternative, based on AES-CTR mode (#12636)

gosubpl committed:
another alternative, based on AES-CTR mode (#12636)

gosubpl commented:
Copying the comment here, as it is not shown by default for the "outdated" block.

I have added `AESCounterBuiltinCTRRNG` that is using `AES/CTR/NoPadding`. To allow verification I have added custom iv with value of 1 but, obviously, this can be removed before merge.
Works great, only issue being that it is even slower, as the zero block needs now to be XORed with the encrypted counter. So the performance looks like that:
```
[info] Benchmark                          Mode  Cnt    Score    Error  Units
[info] RandomReplacementBenchmark.runCtr    ss   15  299.485 ± 23.466  ms/op
[info] RandomReplacementBenchmark.runNew    ss   15  238.772 ± 12.127  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  215.453 ± 24.637  ms/op

[info] RandomReplacementBenchmark.runCtr    ss   15  283.690 ± 17.258  ms/op
[info] RandomReplacementBenchmark.runNew    ss   15  251.412 ± 25.066  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  231.133 ± 29.139  ms/op

[info] RandomReplacementBenchmark.runCtr    ss   15  314.895 ± 45.711  ms/op
[info] RandomReplacementBenchmark.runNew    ss   15  250.495 ± 15.762  ms/op
[info] RandomReplacementBenchmark.runOld    ss   15  215.201 ± 23.651  ms/op
```
which means that the CTR algo is **1.5x** slower than the original one. If that is fine, great, I will happily replace that as hand-built counter is also hurting my eyes.
Of course, aye to reseeding. This is missing from the original implementation, and, for realistic messaging volumes, badly needed. I will add this as soon as we settle the algorithm.

dragos committed:
Update package declaration to fit directory.

patriknw committed:
improve streams actor integration docs, #21698 (#21700)

* mapAsync + ask should be the first choice
* add missing Source.queue
* prefer actorRefWithAck
* move ActorPublisher and ActorSubscriber to the end
  with additional warning
* fix wrong doc of SourceQueue offer
* and add missing java api

Commit:
Rewrite Sink.ignore as a GraphStage #21527

skisel committed:
Test case for persistence plugins re large snapshots #21704

2m committed:
Add apiURL setting for cross-linking scaladoc in projects depending on Akka (#21734)

bantonsson committed:
Make the akka-scala-nightly project work as expected

* Added new scala-java8-compat version for Scala 2.12.0-RC2+
* Fixed a diff between ProjectRef and Project that never matched

ktoso committed:
+slf,act #21671 allow using Markers with LoggingAdapter (#21707)

* +slf,act #21671 allow using Markers with LoggingAdapter

* +slf,act #21671 allow using Markers with LoggingAdapter

* improve compatibility

* add docs

* address review comments

* actually print the marker

johanandren committed:
New EventEnvelope including Offset rather than Long #21722

bantonsson committed:
Merge pull request #21737 from bantonsson/make-akka-scala-nightly-great-again

Make the akka-scala-nightly project work as expected

johanandren commented:
Relevant for this PR, is the comment Will made on the SSL config doc ticket: https://github.com/akka/akka/pull/21730#discussion_r85265351

gosubpl committed:
added performance comparison with SecureRandom (#12636)

gosubpl committed:
added performance comparison with SecureRandom (#12636)

gosubpl commented:
@johanandren yes, it is. Please see performance comparison below, `SecureRandom` is like **12x** slower.
```
[info] Benchmark                             Mode  Cnt     Score     Error  Units
[info] RandomReplacementBenchmark.runCtr       ss   15   297.946 ±  39.956  ms/op
[info] RandomReplacementBenchmark.runNew       ss   15   264.695 ±  35.462  ms/op
[info] RandomReplacementBenchmark.runOld       ss   15   275.792 ±  63.000  ms/op
[info] RandomReplacementBenchmark.runSecRnd    ss   15  3143.013 ± 378.316  ms/op

[info] Benchmark                             Mode  Cnt     Score     Error  Units
[info] RandomReplacementBenchmark.runCtr       ss   15   335.211 ±  77.456  ms/op
[info] RandomReplacementBenchmark.runNew       ss   15   261.172 ±  48.231  ms/op
[info] RandomReplacementBenchmark.runOld       ss   15   246.999 ±  36.847  ms/op
[info] RandomReplacementBenchmark.runSecRnd    ss   15  3172.568 ± 449.505  ms/op
```

I believe that when I first suggested in akka/dev gitter channel to just go with `SecureRandom`, @patriknw opposed, saying (if I correctly recall) that we need something faster. If that is the case, from this comparison I would lean towards @drewhk suggestion to use straight `AES-CTR` mode, as some of its relative slowness would probably come from additional xoring and (hopefully) not ending the "add one to CTR" when there is nothing more to carry which could be susceptible to timing attacks. Other options would be [Fortuna](https://en.wikipedia.org/wiki/Fortuna_(PRNG)) - which practically is AES-CTR with rekeying, but it would have to be probably well tested according to [NIST SP 800-22](http://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-22r1a.pdf) to avoid errors in the implementation (I think that uncommons.math's PRNGs were statistically tested).
Basically if you think `SecureRandom` (straight /dev/urandom, according to @wsargent post) is the way to go, all PRNGs from `uncommons.math` should be deprecated and this PR close. Then, at the next opportunity the whole library can be removed. Otherwise, please let me know which algo you would prefer to implement.

wsargent commented:
Okay, that's interesting.  It makes sense that even a straight call to `/dev/urandom` is slower than a CSPRNG.  It looks like the benchmark is directly against the algorithm though -- is there a comparative test with JSSE with different RNG?

gosubpl commented:
@wsargent will try to cobble something up, as I am also curious what about other RNGs :)

wsargent commented:
Could this be related to https://bugs.openjdk.java.net/browse/JDK-8098581 ?

gosubpl commented:
Hmmm... maybe, but I wouldn't exactly call my request _small sized_:
```
inputSecRnd = Array.fill[Byte](10000000)(1)
```
My way of thinking was that PRNGs are faster than "real" RNGs due to high level of re-use of the original randomness. Need constant re-seeding, though, as the entropy decays.

jrudolph committed:
=rem #13874 cleanup SSLSettings (#21746)

patriknw committed:
Merge pull request #21727 from dragos/issue/misplaced-file

Update package declaration to fit directory.

rkuhn committed:
Scala 2.12 does not name mangle private fields (#21742)

When we can emit a private field, no need to mangle its name

https://github.com/scala/community-builds/pull/317\#issuecomment-254931492

patriknw committed:
send terminationMessage to singleton when leaving last, #21592 (#21654)

* send terminationMessage to singleton when leaving last, #21592

* When leaving last node, i.e. no newOldestOption, the manager was
  just stopped. The change is to send the terminationMessage also
  in this case and wait until the singleton actor is terminated
  before stopping the manager.
* Also changed so that the singleton is stopped immediately when
  cluster has been terminated when last node is leaving, i.e.
  no newOldestOption. Previously it retried until maxTakeOverRetries
  before stopping.
* More comprehensive test of this scenario in ClusterSingletonManagerLeaveSpec

* increase test timeout

ktoso committed:
+act,rem,str add SECURITY marker to logs (#21749)

* +act,rem,str add SECURITY marker to logs

* Update TlsSpec.scala

ktoso committed:
+doc #21668 better docs for configuring ssl (#21730)

* +doc #21668 better docs for configuring ssl

* address some comments, need to address the rest

patriknw committed:
small test changes for Scala 2.12 (#21738)

patriknw committed:
Merge pull request #21613 from akka/wip-21194-sharding-min-members-patriknw

shard coordinator should wait until min-members regions registered, #21194

ktoso committed:
+str #18793 StageLogging that allows logger access in stages (#21696)

* +str #18793 StageLogging that allows logger access in stages
Also, non ActorMaterializers can opt-into providing a logger here.

* +str #18794 add javadsl for StageLogging

* fix missing test method on compile only class

jrudolph committed:
=pro merge ssl-config-akka from ssl-config project into akka-stream (#21551)

This will fix the cyclic dependency issue between the ssl-config repo
and akka.

It would have been better if akka-stream would not require
these changes at all but com.typesafe.sslconfig.akka.AkkaSSLConfig is
part of the public interface of akka-stream's TLS stage. So, this
can only be fixed in the next major version.

Source code was copied over from the tree at commit
https://github.com/typesafehub/ssl-config/commit/470fae76f38dbbd4a42a4000675f66fa67e5f1da

See https://github.com/typesafehub/ssl-config/issues/47

ktoso committed:
reference.conf MUST end with newline (#21755)

jrudolph committed:
Add flag to enable mutual certificate authentication for old Akka Remote SSL transport (#21748)

* =rem #13874 further cleanup of SSLSettings / NettySSLSupport

* +rem #13874 allow requiring mutual authentication for old akka remote ssl transport

ktoso committed:
fix mima warnings (#21756)

ktoso committed:
Remove accidentally re-added akka-http files (#21759)

gosubpl committed:
Merge branch 'wip/12636-alternative-random-function' of github.com:gosubpl/akka into wip/12636-alternative-random-function

gosubpl commented:
Closing as I have this mistakenly rebased on recent changes in master. Will reopen again.

gosubpl commented:
Discussion moved to https://github.com/akka/akka/issues/21740 . Please note that the tests in this PR were done on OpenJDK.

drewhk commented:
@wsargent of course urandom is slower, there is lot's of overhead accessing files, even if they are not backed by disc compared to calling into AES directly (which has some hardware acceleration on recent JDKs although not great).
